
From: Mikael Pettersson <mikpe@csd.uu.se>

This patch eliminates the illegal task_lock() perfctr's
inheritance feature introduced in release_task().

- Changed __vperfctr_release() to use schedule_work() to do
  the task_lock(parent) etc in a different thread's context.
  This is because release_task() has a write lock on the
  task list lock, and task_lock() is forbidden in that case.
  When current == parent, this is bypassed and the merge work
  is done immediately without taking task_lock().
  Added children_lock to struct vperfctr, to synchronise
  accesses (release/update_control/read) to the children array.

Signed-off-by: Mikael Pettersson <mikpe@csd.uu.se>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Index: xx-sources/Documentation/perfctr/virtual.txt
===================================================================
--- xx-sources.orig/Documentation/perfctr/virtual.txt	2004-09-01 13:42:40.832744240 -0400
+++ xx-sources/Documentation/perfctr/virtual.txt	2004-09-01 13:42:41.001718552 -0400
@@ -140,13 +140,9 @@
    before accessing the perfctr pointer.
 
 5. release_task().
-   While reaping a child, the kernel only takes the tasklist_lock to
-   prevent the parent from changing or disappearing. This does not
-   prevent the parent's perfctr state pointer from changing. Concurrent
-   accesses to the parent's "children counts" state are also possible.
-
-   To avoid these issues, perfctr_release_task() performs a task_lock()
-   on the parent.
+   Reaping a child may or may not be done by the parent of that child.
+   When done by the parent, no lock is taken. Otherwise, a task_lock()
+   on the parent is done before accessing its thread's perfctr pointer.
 
 The Pseudo File System
 ----------------------
Index: xx-sources/drivers/perfctr/version.h
===================================================================
--- xx-sources.orig/drivers/perfctr/version.h	2004-09-01 13:42:40.637773880 -0400
+++ xx-sources/drivers/perfctr/version.h	2004-09-01 13:42:41.002718400 -0400
@@ -1 +1 @@
-#define VERSION "2.7.4"
+#define VERSION "2.7.5"
Index: xx-sources/drivers/perfctr/virtual.c
===================================================================
--- xx-sources.orig/drivers/perfctr/virtual.c	2004-09-01 13:42:40.639773576 -0400
+++ xx-sources/drivers/perfctr/virtual.c	2004-09-01 13:42:41.004718096 -0400
@@ -42,10 +42,15 @@
 #ifdef CONFIG_PERFCTR_INTERRUPT_SUPPORT
 	unsigned int iresume_cstatus;
 #endif
-	/* protected by task_lock(owner) */
+	/* children_lock protects inheritance_id and children,
+	   when parent is not the one doing release_task() */
+	spinlock_t children_lock;
 	unsigned long long inheritance_id;
 	struct perfctr_sum_ctrs children;
-	struct work_struct free_work;
+	/* schedule_work() data for when an operation cannot be
+	   done in the current context due to locking rules */
+	struct work_struct work;
+	struct task_struct *parent_tsk;
 };
 #define IS_RUNNING(perfctr)	perfctr_cstatus_enabled((perfctr)->cpu_state.cstatus)
 
@@ -171,8 +176,8 @@
 {
 	if (!atomic_dec_and_test(&perfctr->count))
 		return;
-	INIT_WORK(&perfctr->free_work, scheduled_vperfctr_free, perfctr);
-	schedule_work(&perfctr->free_work);
+	INIT_WORK(&perfctr->work, scheduled_vperfctr_free, perfctr);
+	schedule_work(&perfctr->work);
 }
 
 static unsigned long long new_inheritance_id(void)
@@ -366,39 +371,67 @@
  * A task is being released. If it inherited its perfctr settings
  * from its parent, then merge its final counts back into the parent.
  * Then unlink the child's perfctr.
- * PRE: caller holds tasklist_lock.
+ * PRE: caller has write_lock_irq(&tasklist_lock).
  * PREEMPT note: preemption is disabled due to tasklist_lock.
+ *
+ * When current == parent_tsk, the child's counts can be merged
+ * into the parent's immediately. This is the common case.
+ *
+ * When current != parent_tsk, the parent must be task_lock()ed
+ * before its perfctr state can be accessed. task_lock() is illegal
+ * here due to the write_lock_irq(&tasklist_lock) in release_task(),
+ * so the operation is done via schedule_work().
  */
-void __vperfctr_release(struct task_struct *child_tsk)
+static void do_vperfctr_release(struct vperfctr *child_perfctr, struct task_struct *parent_tsk)
 {
-	struct task_struct *parent_tsk;
 	struct vperfctr *parent_perfctr;
-	struct vperfctr *child_perfctr;
 	unsigned int cstatus, nrctrs, i;
 
-	/* Our caller holds tasklist_lock, protecting child_tsk->parent.
-	   We must also task_lock(child_tsk->parent), to prevent its
-	   ->thread.perfctr from changing. */
-	parent_tsk = child_tsk->parent;
-	task_lock(parent_tsk);
 	parent_perfctr = parent_tsk->thread.perfctr;
-	child_perfctr = child_tsk->thread.perfctr;
-	if (parent_perfctr && child_perfctr &&
-	    parent_perfctr->inheritance_id == child_perfctr->inheritance_id) {
-		cstatus = parent_perfctr->cpu_state.cstatus;
-		if (perfctr_cstatus_has_tsc(cstatus))
-			parent_perfctr->children.tsc +=
-				child_perfctr->cpu_state.tsc_sum +
-				child_perfctr->children.tsc;
-		nrctrs = perfctr_cstatus_nrctrs(cstatus);
-		for(i = 0; i < nrctrs; ++i)
-			parent_perfctr->children.pmc[i] +=
-				child_perfctr->cpu_state.pmc[i].sum +
-				child_perfctr->children.pmc[i];
+	if (parent_perfctr && child_perfctr) {
+		spin_lock(&parent_perfctr->children_lock);
+		if (parent_perfctr->inheritance_id == child_perfctr->inheritance_id) {
+			cstatus = parent_perfctr->cpu_state.cstatus;
+			if (perfctr_cstatus_has_tsc(cstatus))
+				parent_perfctr->children.tsc +=
+					child_perfctr->cpu_state.tsc_sum +
+					child_perfctr->children.tsc;
+			nrctrs = perfctr_cstatus_nrctrs(cstatus);
+			for(i = 0; i < nrctrs; ++i)
+				parent_perfctr->children.pmc[i] +=
+					child_perfctr->cpu_state.pmc[i].sum +
+					child_perfctr->children.pmc[i];
+		}
+		spin_unlock(&parent_perfctr->children_lock);
 	}
+	schedule_put_vperfctr(child_perfctr);
+}
+
+static void scheduled_release(void *data)
+{
+	struct vperfctr *child_perfctr = data;
+	struct task_struct *parent_tsk = child_perfctr->parent_tsk;
+
+	task_lock(parent_tsk);
+	do_vperfctr_release(child_perfctr, parent_tsk);
 	task_unlock(parent_tsk);
+	put_task_struct(parent_tsk);
+}
+
+void __vperfctr_release(struct task_struct *child_tsk)
+{
+	struct task_struct *parent_tsk = child_tsk->parent;
+	struct vperfctr *child_perfctr = child_tsk->thread.perfctr;
+
 	child_tsk->thread.perfctr = NULL;
-	schedule_put_vperfctr(child_perfctr);
+	if (parent_tsk == current)
+		do_vperfctr_release(child_perfctr, parent_tsk);
+	else {
+		get_task_struct(parent_tsk);
+		INIT_WORK(&child_perfctr->work, scheduled_release, child_perfctr);
+		child_perfctr->parent_tsk = parent_tsk;
+		schedule_work(&child_perfctr->work);
+	}
 }
 
 /* schedule() --> switch_to() --> .. --> __vperfctr_suspend().
@@ -544,10 +577,10 @@
 		if (!(control->preserve & (1<<i)))
 			perfctr->cpu_state.pmc[i].sum = 0;
 
-	task_lock(tsk);
+	spin_lock(&perfctr->children_lock);
 	perfctr->inheritance_id = new_inheritance_id();
 	memset(&perfctr->children, 0, sizeof perfctr->children);
-	task_unlock(tsk);
+	spin_unlock(&perfctr->children_lock);
 
 	if (tsk == current)
 		vperfctr_resume(perfctr);
@@ -648,10 +681,10 @@
 	}
 	if (childrenp) {
 		if (tsk)
-			task_lock(tsk);
+			spin_lock(&perfctr->children_lock);
 		tmp->children = perfctr->children;
 		if (tsk)
-			task_unlock(tsk);
+			spin_unlock(&perfctr->children_lock);
 	}
 	if (tsk == current)
 		preempt_enable();
